File: AndroidServiceCommunication/app/src/main/java/lib/serviceconnector/ServiceManager.java
package lib.serviceconnector;

import android.app.Activity;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.Handler;
import android.os.IBinder;
import android.os.Message;
import android.os.Messenger;
import android.os.RemoteException;
import android.widget.Toast;

import static lib.serviceconnector.ServiceManager.MessageReceiver.*;

import java.lang.ref.WeakReference;

/**
 * Created by Lenovo7 on 8/23/2015.
 */
public final class ServiceManager {

    private Intent mServiceIntent;
    private WeakReference<DialServiceConnection> mDialServiceConnWfr;
    private Messenger mSenderToService;
    private Activity mActivity;
    private MessageReceiver mMessageReceiver;
    private final Messenger mMessengerReceiver = new Messenger(new ServiceReceiverHandler());

    public interface MessageReceiver {

        int SERVICE_CONNECTED = 1;
        int SERVICE_DISCONNECT = 2;
        int SERVICE_STOP = 3;
        int SERVICE_SENT_DATA = 10;

        void onMessageReceivedFromService(Message message);

        void onMessageReceivedFromService(Object object);
    }

    /***
     * @param activity
     * @param messageReceiver
     */
    public ServiceManager(Activity activity, MessageReceiver messageReceiver) {
        mActivity = activity;
        mMessageReceiver = messageReceiver;
    }

    /***
     * @param className
     */
    public void startService(Class className) {

        mDialServiceConnWfr = new WeakReference<DialServiceConnection>(new DialServiceConnection());
        mServiceIntent = new Intent(mActivity, className);
        mActivity.startService(mServiceIntent);
        mActivity.bindService(mServiceIntent, mDialServiceConnWfr.get(), Context.BIND_AUTO_CREATE);
    }

    /***
     * @param message
     */
    public void sendMessageToService(Message message) {
        try {
            if (mSenderToService != null) {
                message.replyTo = mMessengerReceiver;
                mSenderToService.send(message);
            } else {
                Toast.makeText(mActivity, "Sender is undefined", Toast.LENGTH_SHORT).show();
            }
        } catch (RemoteException e) {
            e.printStackTrace();
            Toast.makeText(mActivity, e.getMessage(), Toast.LENGTH_SHORT).show();
        }
    }

    /***
     * @param message
     */
    public void sendMessageToService(String message) {
        try {
            if (mSenderToService != null) {
                Message msg = new Message();
                msg.obj = message;
                msg.replyTo = mMessengerReceiver;
                mSenderToService.send(msg);
            } else {
                Toast.makeText(mActivity, "Sender is undefined", Toast.LENGTH_SHORT).show();
            }
        } catch (RemoteException e) {
            e.printStackTrace();
            Toast.makeText(mActivity, e.getMessage(), Toast.LENGTH_SHORT).show();
        }
    }

    private class DialServiceConnection implements ServiceConnection {

        public void onServiceConnected(ComponentName className, IBinder service) {
            mSenderToService = new Messenger(service);
            Message msg = Message.obtain(null, SERVICE_CONNECTED);
            sendMessageToService(msg);
        }

        /***
         * @param className
         */
        public void onServiceDisconnected(ComponentName className) {
            //Toast.makeText(mActivity.get, "SERVICE_CONNECTED", Toast.LENGTH_SHORT).show();
        }
    }

    private class ServiceReceiverHandler extends Handler {
        @Override
        public void handleMessage(Message msg) {
            super.handleMessage(msg);
            if (msg != null) {
                switch (msg.what) {
                    case SERVICE_SENT_DATA:
                        if (mMessageReceiver != null) {
                            mMessageReceiver.onMessageReceivedFromService(msg.obj);
                        }
                        break;
                    case SERVICE_DISCONNECT:
                        disconnectedToService();
                        break;
                    case SERVICE_STOP:
                        stopService();
                        break;
                    default:
                        if (mMessageReceiver != null) {
                            mMessageReceiver.onMessageReceivedFromService(msg);
                        }
                        break;
                }
            }
        }
    }

    public void disconnectedToService() {
        try {
            sendMessageToService(Message.obtain(null, SERVICE_DISCONNECT));
            mActivity.unbindService(mDialServiceConnWfr.get());
            mDialServiceConnWfr.clear();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void stopService() {
        try {
            sendMessageToService(Message.obtain(null, SERVICE_STOP));
            mActivity.unbindService(mDialServiceConnWfr.get());
            mActivity.stopService(mServiceIntent);
            mDialServiceConnWfr.clear();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: AndroidServiceCommunication/app/src/main/java/lib/connector/ServiceManager.kt
package lib.connector

import android.app.Activity
import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.*
import android.widget.Toast
import java.lang.ref.WeakReference

/**
 * Manage the whole Activity-Service Communication
 * @property activity Activity
 * @property messageReceiver MessageReceiver?
 * @property appServiceConnection WeakReference<AppServiceConnection>?
 * @property serviceIntent Intent?
 * @property senderToService Messenger?
 * @property messengerReceiver Messenger
 * @constructor
 */
class ServiceManager(
    private val activity: Activity,
    private val messageReceiver: MessageReceiver?
) : HandleMessageCallback {

    private var appServiceConnection: WeakReference<AppServiceConnection>? = null
    private var serviceIntent: Intent? = null
    private var senderToService: Messenger? = null
    private var messengerReceiver: Messenger = Messenger(ServiceHandler(this))// = Messenger(this)

    /***
     * @param className
     */
    fun startService(className: Class<*>) {
        appServiceConnection = WeakReference(AppServiceConnection())
        serviceIntent = Intent(activity, className)
        activity.startService(serviceIntent)
        activity.bindService(serviceIntent, appServiceConnection?.get()!!, Context.BIND_AUTO_CREATE)
    }

    /***
     * @param message
     */
    fun sendMessageToService(message: String) {
        try {
            senderToService?.let {
                val msg = Message()
                msg.obj = message
                msg.replyTo = messengerReceiver
                it.send(msg)
            } ?: run {
                Toast.makeText(
                    activity.applicationContext,
                    "Sender is undefined", Toast.LENGTH_LONG
                ).show()
            }
        } catch (e: RemoteException) {
            e.printStackTrace()
            Toast.makeText(activity.applicationContext, e.message, Toast.LENGTH_LONG).show()
        }
    }

    /***
     * @param message
     */
    fun sendMessageToService(message: Message) {
        try {
            senderToService?.let {
                message.replyTo = messengerReceiver
                it.send(message)
            } ?: run {
                Toast.makeText(
                    activity.applicationContext,
                    "Sender is undefined", Toast.LENGTH_LONG
                ).show()
            }
        } catch (e: RemoteException) {
            e.printStackTrace()
            Toast.makeText(activity.applicationContext, e.message, Toast.LENGTH_LONG).show()
        }
    }

    fun disconnectedToService() {
        try {
            sendMessageToService(Message.obtain(null, MessageReceiver.SERVICE_DISCONNECT))
            activity.unbindService(appServiceConnection?.get()!!)
            appServiceConnection?.clear()
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    private fun stopService() {
        try {
            sendMessageToService(Message.obtain(null, MessageReceiver.SERVICE_STOP))
            activity.unbindService(appServiceConnection?.get()!!)
            activity.stopService(serviceIntent)
            appServiceConnection?.clear()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun onMessageReceivedFomHandler(message: Message) {
        when (message.what) {
            MessageReceiver.SERVICE_SENT_DATA -> {
                messageReceiver?.onMessageReceivedFromService(message.obj)
            }
            MessageReceiver.SERVICE_DISCONNECT -> {
                disconnectedToService()
            }
            MessageReceiver.SERVICE_STOP -> {
                stopService()
            }
            else -> {
                messageReceiver?.onMessageReceivedFromService(message)
            }
        }
    }

    private inner class AppServiceConnection : ServiceConnection {

        override fun onServiceDisconnected(name: ComponentName?) {
        }

        override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
            senderToService = Messenger(service)
            val msg = Message.obtain(null, MessageReceiver.SERVICE_CONNECTED)
            sendMessageToService(msg)
        }
    }
}

